test/correctness/device_crop.cpp

```diff
@@ -0,0 +1,136 @@
+        assert(gpu_buf.raw_buffer()->device_interface != nullptr);
+
+        gpu_buf.copy_to_host();
+        for (int i = 0; i < 64; i++) {
+            for (int j = 0; j < 64; j++) {
+              assert(gpu_buf(32 + i, 32 + j) == (i + 32) + 256 * (j + 32));
+            }
+        }
+    }
+
+    printf("Test nondestructive cropping.\n");
+    {
+        Halide::Runtime::Buffer<int32_t> gpu_buf = make_gpu_buffer();
+        assert(gpu_buf.raw_buffer()->device_interface != nullptr);
+
+        Halide::Runtime::Buffer<int32_t> cropped = gpu_buf.cropped({ {32, 64} , {32, 64} });
+        assert(cropped.raw_buffer()->device_interface != nullptr);
+
+        cropped.copy_to_host();
+        for (int i = 0; i < 64; i++) {
+            for (int j = 0; j < 64; j++) {
+                assert(cropped(32 + i, 32 + j) == (i + 32) + 256 * (j + 32));
+            }
+        }
+    }
+
+    printf("Test crop of a crop\n");
+    {
+        Halide::Runtime::Buffer<int32_t> gpu_buf = make_gpu_buffer();
+        assert(gpu_buf.raw_buffer()->device_interface != nullptr);
+
+        Halide::Runtime::Buffer<int32_t> cropped = gpu_buf.cropped({ {32, 64} , {32, 64} });
+        assert(cropped.raw_buffer()->device_interface != nullptr);
+
+        Halide::Runtime::Buffer<int32_t> cropped2 = cropped.cropped({ {40, 16} , {40, 16} });
+        assert(cropped2.raw_buffer()->device_interface != nullptr);
+
+        cropped.copy_to_host();
+        for (int i = 0; i < 64; i++) {
+            for (int j = 0; j < 64; j++) {
+                assert(cropped(32 + i, 32 + j) == (i + 32) + 256 * (j + 32));
+            }
+        }
+
+        cropped2.copy_to_host();
+        for (int i = 0; i < 16; i++) {
+            for (int j = 0; j < 16; j++) {
+                assert(cropped2(40 + i, 40 + j) == (i + 40) + 256 * (j + 40));
+            }
+        }
+    }
+    
+    printf("Test parent going out of scope before crop.\n");
+    {
+        Halide::Runtime::Buffer<int32_t> cropped;
+
+        {
+            Halide::Runtime::Buffer<int32_t> gpu_buf = make_gpu_buffer();
+            assert(gpu_buf.raw_buffer()->device_interface != nullptr);
+
+            cropped = gpu_buf.cropped({ {32, 64} , {32, 64} });
+            assert(cropped.raw_buffer()->device_interface != nullptr);
+        }
+
+        cropped.copy_to_host();
+        for (int i = 0; i < 64; i++) {
+            for (int j = 0; j < 64; j++) {
+                assert(cropped(32 + i, 32 + j) == (i + 32) + 256 * (j + 32));
+            }
+        }
+    }
+
+    printf("Test realizing to/from crop.\n");
+    {
+        ImageParam in(Int(32), 2);
+        Var x, y;
+        Func f;
+        f(x, y) = in(x, y) + 42;
+
+        Var xi, yi;
+        f.gpu_tile(x, y, xi, yi, 8, 8);
+
+        Halide::Buffer<int32_t> gpu_input = make_gpu_buffer();
+        Halide::Buffer<int32_t> gpu_output = make_gpu_buffer();
+
+        // TODO(steven-johnson|abadams): Why doesn't the forward forward?
```

Comment:
The issue seems to be that the perfect-forwarding approach can't handle the overloads of Runtime::Buffer::crop(); explicitly tagging this as std::vector<std::pair<int, int>> allows it to compile (but that's not really acceptable). We could fix this by adding explicit forwarding code to Buffer, but maybe there's a way to finesse it to have the forwarding code work.

Reply:
Yeah, the short answer is: "perfect forwarding" isn't perfect; most notably, it can't handle initializer lists. That said, I think I can add an overload to work around it; look for separate PR.